src/Parse.idr

```diff
@@ -83,12 +93,13 @@ tdef =
               )
               (parens (and (guardM (\(mp, nam) => pushName nam <$> lookup nam mp) $ mand (lift get) alphas)
                            (map {a=Parser' _} (nelist . ignoreSpaces) rec)))
-        , cmap (Z ** T0) $ string "0"
-        , cmap (Z ** T1) $ string "1"
+        , 
+          cmap (Z ** T0) $ char '0'
+        , cmap (Z ** T1) $ char '1'
         , nary rec '*' TProd
         , nary rec '+' TSum
         , map (\n => (S n ** TVar $ last {n})) $
-            parens (rand (ignoreSpaces (string "var")) (ignoreSpaces decimalNat))
+            parens $ rand (ignoreSpaces $ string "var") (ignoreSpaces $ commit $ decimalNat)
```

Comment:
What does `commit` do?

Reply:
Internally it turns soft fails into hard ones, limiting backtracking. This is essentially what allows error positions to be more or less specific, otherwise you'll just report the root of the whole offending tree.

Reply:
Ah I see, thanks :)